app/gui/src/controller/graph/widgets.rs

```diff
@@ -0,0 +1,481 @@
+                suggestion.arguments.iter().map(|arg| arg.name.clone().into()).collect();
+            visualization_modified = true;
+        }
+
+        self.node_id = req.node_id;
+        self.call_expression = req.call_expression;
+        visualization_modified
+    }
+
+    fn last_updates(&self) -> Option<(NodeId, WidgetUpdates)> {
+        self.last_updates.as_ref().map(|updates| {
+            let call_id = self.call_expression;
+            let updates = WidgetUpdates { call_id, updates: updates.clone() };
+            (self.node_id, updates)
+        })
+    }
+
+    fn request_visualization(&mut self, manager: &Rc<Manager>, target_expression: ast::Id) {
+        // When visualization is requested, remove stale queried value to prevent updates  while
+        // language server request is pending.
+        self.last_updates.take();
+        let vis_metadata = self.visualization_metadata();
+        manager.request_visualization(target_expression, vis_metadata);
+    }
+
+    /// Generate visualization metadata for this query.
+    fn visualization_metadata(&self) -> Metadata {
+        let relevant_arguments = self.arguments.split_first().map_or_default(|(_self, args)| args);
+        let arguments: Vec<Code> = vec![
+            Self::escape_visualization_argument(&self.method_name).into(),
+            Self::arg_sequence(relevant_arguments).into(),
+        ];
+
+        let preprocessor = visualization::instance::PreprocessorConfiguration {
+            module:    WIDGET_VISUALIZATION_MODULE.into(),
+            method:    WIDGET_VISUALIZATION_METHOD.into(),
+            arguments: Rc::new(arguments),
+        };
+        Metadata { preprocessor }
+    }
+
+    /// Escape a string to be used as a visualization argument. Transforms the string into an enso
+    /// expression with string literal.
+    fn escape_visualization_argument(arg: &str) -> String {
+        let segment = ast::SegmentPlain { value: arg.into() };
+        let text = ast::TextLineRaw { text: vec![segment.into()] };
+        text.repr()
+    }
+
+    /// Escape a list of strings to be used as a visualization argument. Transforms the strings into
+    /// an enso expression with a list of string literals.
+    fn arg_sequence(args: &[ImString]) -> String {
+        let mut buffer = String::from("[");
+        for (i, arg) in args.iter().enumerate() {
+            if i > 0 {
+                buffer.push_str(", ");
+            }
+            buffer.push_str(&Self::escape_visualization_argument(arg));
+        }
+        buffer.push(']');
+        buffer
+    }
+}
+
+
+
+/// ===============================
+/// === WidgetVisualizationData ===
+/// ===============================
+
+/// A type representing the data received from the widget visualization for a single widget.
+///
+/// The structure of this struct is dictated by the expected widget visualization JSON result shape.
+#[derive(Debug, serde::Deserialize)]
+struct WidgetVisualizationData {
```

Comment:
We are in `widget` module, you don't need to repeat the `Widget` name here. It should be used as `widget::VisualizationData`.